design/metastore.md

```diff
@@ -0,0 +1,102 @@
+# Next Generation Metastore
+
+### Overview
+
+Deliver manageability of metastore information alongside the object-store data as part of lakeFS.
+
+### Goals
+
+- Metastore entities are versioned: using the same commit and branching abilities that lakeFS provides
+- Metastore entities are versioned together with the data: an atomic merge operation should contain both data mutations as well as metastore mutations.
+- Metastore entities are diff-able: can view the changes made to the metadata between commits
+- Maintain some compatibility with existing HMS, or at least a narrow set of its versions and API
+- Extend functionality to also support next-generation table formats in addition to hive-style tables (Notably, Iceberg and Delta Lake) - long term goal.
+
+
+### Non-Goals
+
+- Support storing metadata inside existing Hive Metastore implementations
+- Provide any significant change in performance or throughput of metadata operations
+- Provide compatibility with other specific versions of Hive or Hive Metastore
+
+
+### Design Diagram
+
+[lakeFS Metastore](diagrams/metastore.png)
+
+In this diagram:
+1. Spark/Hive/Trino - data and metadata clients
+1. lakeFS Metastore Thrift Proxy - act as the client's metastore by implementing the Hive Metastore Thrift protocol
+1. lakeFS - Include the new following components:
+   - Additional operations on OpenAPI that will 
+   - Metadata catalog that will wrap Graveler to serve metadata like we do for the object store.
+
+The diagram show how Spark/Hive/Trino connects to lakeFS's metastore proxy (which acts as Hive Metastore).
+The request to the metastore proxy will invoke the equivalent operation on lakeFS's OpenAPI.
+Inside lakeFS API implementation will use the metastore catalog, that is implemented by Graveler as the storage engine.
+
+
+### Data Model
+
+Using Gravler we can implement a Metastore catalog for all the metadata properties.
```

Comment:
Agreed.  The obvious alternative is to keep the metastore catalog as "normal" files with "special" names inside the repo.  That makes _some_ things much easier (e.g. viewing history, accessing from a client, etc.).  Other things might become harder to do.

Let's explain here why we're adding HMS as more metadata rather than as more data.

There are no advantages that I can see in terms of using Graveler to manage huge files -- I do not believe catalogs can be that big.

The obvious place where storing inside a Graveler is an advantage is *if* merges become trivial.  Then again, when merges are trivial then merging a single structured file should also be quite easy.